### ISSUE TITLE
model does not converge when training, the loss is nan

### ISSUE THREAD

**Issue by cherry-licongyi** [ssg-research/WAFFLE#8]
### Description
There is no loss value in the training process, and the model does not converge. I did not change the code, only changed the epoch in the configuration file.

It is the same results when i run  `python3 main.py --config_file configurations/baseline_experiments/cifar10/1.ini --experiment training`  and `python3 main.py --config_file configurations/baseline_experiments/mnist/1.ini --experiment training` 

### code
```
client_model.train()
                correct = 0
                train_loss = 0
                # Create optimizer for wrokers
                if self.pretrain_epochs == 150 or self.model_architecture == 'vgg16': #use smaller lr in random noise
                    client_optimizer = torch.optim.SGD(params=client_model.parameters(), lr=0.01)
                else:
                    client_optimizer = torch.optim.SGD(params=client_model.parameters(), lr=0.1) #0.1

                for local_epoch in range(self.training_ops.local_epoch):
                    for data, target in federated_train_loader[client_i]:
                        # print(data.location)
                        if self.use_cuda:
                            data, target = data.cuda(), target.cuda() 
                        client_optimizer.zero_grad()
                        output = client_model(data)

                        if self.model_architecture == "MNIST_L5":
                            loss = nn.functional.nll_loss(output, target)
                        else:
                            loss = nn.CrossEntropyLoss()(output, target)
                        loss.backward()
                        client_optimizer.step()
                        loss = loss.get() # get the client's loss back
                        
                        train_loss += loss.item()
                        if local_epoch == self.training_ops.local_epoch-1:
                            with torch.set_grad_enabled(False):
                                output = client_model(data)
                                _, pred = torch.max(output.data, 1)
                                pred = pred.get()
                                target = target.get()
                                correct += pred.eq(target.view_as(pred)).sum().item()
                    train_accuracies[client_i] = 100 * correct / len(federated_train_loader[client_i].dataset) 
                    train_losses[client_i] += train_loss / (len(federated_train_loader[client_i].dataset) * self.training_ops.local_epoch)

                # Print GPU usage to check if there is a memory leak
                print('GPU usage {}'.format(torch.cuda.memory_allocated()))  
                # Move function send client model parameters to aggregator (secure worker)
                client_model.move(secure_worker)         
                updated_w = updateWeight.calculate_weight(updated_w, self.real_client_num, self.client_num, client_model)
                # Increase communication in bytes with the size of updated_w ...
                # ... and multiply it with 2, since download+upload is the total communication 
                Mb += total_size(updated_w)#sys.getsizeof(updated_w)
                print('Worker:{} Train Epoch: {}, worker data size {}, worker avg loss: {:.3f}, worker train acc:{:.2f}'.format(client_i, 
                                                           epoch+1, 
                                                           len(federated_train_loader[client_i].dataset), 
                                                           train_losses[client_i], train_accuracies[client_i]))
                                                           
                with open('./train_record/new_epoch_logs_' + self.model_name + '.txt', 'a+') as file:
                    file.write('Worker:{} Train Epoch: {}, worker data size {}, worker avg loss: {:.3f}, worker train acc:{:.2f}\n'.format(client_i, 
                                                           epoch+1, 
                                                           len(federated_train_loader[client_i].dataset), 
                                                           train_losses[client_i], train_accuracies[client_i]))

                del client_model
                torch.cuda.empty_cache()

            # Receive data in bytes from secure worker to central server
            Mb += total_size(updated_w)
            # Update weights of federated model with average weights
            federated_model = updateWeight.send_weight(federated_model, updated_w)
            # Average loss of workers 
            avg_loss = np.sum(train_losses)/self.client_num
            scores["loss"].append(score.FloatScore(avg_loss))
            # Calculate watermark score after each epoch
            watermark_score = self.test_watermark(federated_model) 
            test_average, test_per_class = self.test_model(federated_model)  
            with open('./train_record/new_epoch_logs_' + self.model_name + '.txt', 'a+') as file:
                file.write('Global model, Epoch:{}, Test Acc before watermark retrain: {}\n'.format(epoch+1, test_average.value))
                file.write('Global model, Epoch:{}, Watermark Acc before watermark retrain: {}\n'.format(epoch+1, watermark_score.value))
```

### output log
```
Global model, Epoch:25, Test Acc before watermark retrain: 10.0

Global model, Epoch:25, Watermark Acc before watermark retrain: 10.0

Testing at 25th epoch, train loss:nan
>>> INFO: Test accuracy: (1000/10000) 10.00%
>>> INFO: Accuracy of the network on the 10000 test images (per-class):
Accuracy of     0 : 99 %
Accuracy of     1 :  0 %
Accuracy of     2 :  0 %
Accuracy of     3 :  0 %
Accuracy of     4 :  0 %
Accuracy of     5 :  0 %
Accuracy of     6 :  0 %
Accuracy of     7 :  0 %
Accuracy of     8 :  0 %
Accuracy of     9 :  0 %
>>> INFO: Finished training. The total communication is: 436800, the total wm retrain rounds are: 0
```

**Comment by cherry-licongyi** [comment]
It looks like there was a problem in the fedavg process， this is the mnist traing log
```
begin time: 1692957305.7994812, file name: configurations/baseline_experiments/mnist/1.ini
Worker:2 Train Epoch: 1, worker data size 12000, worker avg loss: 0.023, worker train acc:99.88
Worker:3 Train Epoch: 1, worker data size 12000, worker avg loss: 0.022, worker train acc:99.87
Worker:0 Train Epoch: 1, worker data size 12000, worker avg loss: 0.023, worker train acc:99.92
Worker:1 Train Epoch: 1, worker data size 12000, worker avg loss: 0.023, worker train acc:99.88
Worker:4 Train Epoch: 1, worker data size 12000, worker avg loss: 0.023, worker train acc:99.85
Global model, Epoch:1, Test Acc before watermark retrain: 98.09
Global model, Epoch:1, Watermark Acc before watermark retrain: 6.0
Global model, Epoch:1, Average_loss: 0.023, Test Acc: 98.09
Worker:2 Train Epoch: 2, worker data size 12000, worker avg loss: nan, worker train acc:10.32
Worker:4 Train Epoch: 2, worker data size 12000, worker avg loss: nan, worker train acc:9.53
Worker:1 Train Epoch: 2, worker data size 12000, worker avg loss: nan, worker train acc:9.80
Worker:0 Train Epoch: 2, worker data size 12000, worker avg loss: nan, worker train acc:10.01
Worker:3 Train Epoch: 2, worker data size 12000, worker avg loss: nan, worker train acc:9.70
Global model, Epoch:2, Test Acc before watermark retrain: 9.8
Global model, Epoch:2, Watermark Acc before watermark retrain: 10.0
Global model, Epoch:2, Average_loss: nan, Test Acc: 9.80
```

**Comment by bussfromspace** [comment]
Hi,

Did you pull the latest updates? I have changed the updateWeight.py file last week, since weight averaging was working wrong when the number clients joining a single aggregation round is different from 10.

**Comment by cherry-licongyi** [comment]
More info during training
```
federated_model param
torch.Size([10, 200]) tensor([ 0.7954,  0.7141, -1.3912, -1.3904,  0.5688,  ..., 
         1.6755, -1.5621,  1.4071,  1.6722, -0.4808,  1.0526,  0.0951, -0.8981])
torch.Size([10]) tensor(0.6569)
>>> INFO: Accuracy of the network on the 100 watermark images (6/100): 6.0
>>> INFO: Test accuracy: (7952/10000) 79.52%
>>> INFO: Accuracy of the network on the 10000 test images (per-class):
Accuracy of     0 : 89 %
Accuracy of     1 : 94 %
Accuracy of     2 : 82 %
Accuracy of     3 : 56 %
Accuracy of     4 : 67 %
Accuracy of     5 : 80 %
Accuracy of     6 : 73 %
Accuracy of     7 : 82 %
Accuracy of     8 : 91 %
Accuracy of     9 : 74 %
Testing at 2th epoch, train loss:0.17316190056564906
>>> INFO: Test accuracy: (7952/10000) 79.52%
>>> INFO: Accuracy of the network on the 10000 test images (per-class):
Accuracy of     0 : 89 %
Accuracy of     1 : 94 %
Accuracy of     2 : 82 %
Accuracy of     3 : 56 %
Accuracy of     4 : 67 %
Accuracy of     5 : 80 %
Accuracy of     6 : 73 %
Accuracy of     7 : 82 %
Accuracy of     8 : 91 %
Accuracy of     9 : 74 %
* Epoch 2/10
* Client 4/5
* local_epoch 0/2
4 train_accuracy 0.0
4 train_loss nan
* local_epoch 1/2
4 train_accuracy 10.316666666666666
4 train_loss nan
GPU usage 203006976
Worker:4 Train Epoch: 3, worker data size 12000, worker avg loss: nan, worker train acc:10.32
* Client 0/5
* local_epoch 0/2
0 train_accuracy 0.0
0 train_loss nan
* local_epoch 1/2
0 train_accuracy 9.7
0 train_loss nan
GPU usage 206234112
Worker:0 Train Epoch: 3, worker data size 12000, worker avg loss: nan, worker train acc:9.70
* Client 3/5
* local_epoch 0/2
3 train_accuracy 0.0
3 train_loss nan
* local_epoch 1/2
3 train_accuracy 9.8
3 train_loss nan
GPU usage 206234112
Worker:3 Train Epoch: 3, worker data size 12000, worker avg loss: nan, worker train acc:9.80
* Client 2/5
* local_epoch 0/2
2 train_accuracy 0.0
2 train_loss nan
* local_epoch 1/2
2 train_accuracy 9.533333333333333
2 train_loss nan
GPU usage 206234112
Worker:2 Train Epoch: 3, worker data size 12000, worker avg loss: nan, worker train acc:9.53
* Client 1/5
* local_epoch 0/2
1 train_accuracy 0.0
1 train_loss nan
* local_epoch 1/2
1 train_accuracy 10.008333333333333
1 train_loss nan
GPU usage 206234112
Worker:1 Train Epoch: 3, worker data size 12000, worker avg loss: nan, worker train acc:10.01
federated_model param
torch.Size([10, 200]) tensor([nan, nan, nan, nan,...,
        nan, nan, nan, nan, nan, nan, nan, nan])
```

**Comment by cherry-licongyi** [comment]
Thank you！
It works now. Really appreciate your help
